collab-document/src/document.rs

```diff
@@ -319,6 +323,34 @@ impl Document {
       .set_block_with_txn(txn, block_id, Some(block.data), Some(&new_parent.id))
   }
 
+  pub fn redo(&self) -> bool {
+    let can_redo = self.can_redo();
+    if can_redo {
+      if let Ok(redo) = self.inner.lock().redo() {
+        return redo;
+      }
+    }
+    false
+  }
+
+  pub fn undo(&self) -> bool {
+    let can_undo = self.can_undo();
+    if can_undo {
```

Comment:
How about 
'''rust
if self.can_undo() {
   self.inner.lock().undo().unwrap_or_default()
} else {
  false
}
'''

Reply:
Actually using try_lock would be better

Reply:
Is the `inner.lock()` have any problem?

Reply:
If the lock takes a long time to process something, it might block others who want to acquire this lock. However, 'try_lock' returns immediately. If undo or redo fails, the user can attempt it again.